#!/usr/bin/env perl
###################################
# Author: Jiang Li
# Email: riverlee2008@gmail.com
# Date: Wed Aug 28 13:23:12 2013
###################################
use strict;
use warnings;

=head data
Download the transcript factors and DNase binding peaks accross cells, 
details are avilalble at http://genome.ucsc.edu/ENCODE/

UCSC has released a new browser track containing 690 datasets of transcription factor ChIP-seq peaks based on data from all five ENCODE TFBS ChIP-seq production groups from the project inception in 2007 through the ENCODE March 2012 data freeze. The track covers 161 unique regulatory factors (generic and sequence-specific factors), spanning 91 human cell types, some under various treatment conditions.

Browser track: Transcription Factor ChIP-seq Uniform Uniform Peaks from ENCODE/Analysis 
File downloads: Directory File selection tool

This track represents peak calls (regions of enrichment) generated by the ENCODE Analysis Working Group (AWG) using the uniform processing pipeline developed for the ENCODE Integrative Analysis effort and published in a set of coordinated papers in September 2012. Peak calls from that effort (based on datasets from the January 2011 ENCODE data freeze) are available at the ENCODE Analysis Data Hub. The new Uniform TFBS track at UCSC includes newer data, slightly modified processing methods, and improved metadata. Quality metrics are included in metadata, with detailed metrics in a quality spreadsheet linked to the track description. Browser users will see the uniform peaks first when using track search for TFBS, and this track is now the default track shown when the ENCODE TF Binding menu item is selected in the browser.

The primary and lab-processed data (along with methods descriptions, credits and references) on which this track is based are available in the following ENCODE tracks: HAIB TFBS, SYDH TFBS, UChicago TFBS, UTA TFBS, UW CTCF Binding. Many thanks to Anshul Kundaje of the ENCODE AWG for providing the uniform peaks data, description, and quality spreadsheets.

=cut

my $datadir="data";

mkdir $datadir if (! -d $datadir);
chdir $datadir or die $!;

my $baseurl="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeAwgTfbsUniform/";

# First download the files.txt
my $url="$baseurl/files.txt";
download($url,"TF_files.txt");

open(IN,"TF_files.txt") or die $!;
while(<IN>){
    my ($file,@others) = split /\t|;/;
    next if (scalar(@others)<4); # remove this line wgEncodeAwgTfbsHaibPfsk1Pol24h8V0416101UniPk.narrowPeak.gz
    download("$baseurl/$file",$file);
}
close IN;

# Download DNase
$baseurl="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeAwgDnaseUniform/";

# First download the files.txt
$url="$baseurl/files.txt";
download($url,"DNase_files.txt");

open(IN,"DNase_files.txt") or die $!;
while(<IN>){
    my ($file,@others) = split /\t|;/;
    next if (scalar(@others)<4); # remove this line wgEncodeAwgTfbsHaibPfsk1Pol24h8V0416101UniPk.narrowPeak.gz
    download("$baseurl/$file",$file);
}
close IN;

## Download the histone mark data
# First download the files.txt
$baseurl="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHistone/";
$url="$baseurl/files.txt";
download($url,"BroadHistone_files.txt");

open(IN,"BroadHistone_files.txt") or die $!;
while(<IN>){
    my ($file,@others) = split /\t|;/;
    next if ($file!~/Peak.gz$/);
    download("$baseurl/$file",$file);
}
close IN;


# merge together
`cat TF_files.txt DNase_files.txt BroadHistone_files.txt >files.txt`;

sub download{
    my ($url,$output) = @_;
    info("Downloading $output");
    my $com="curl $url -o $output";
    `$com`;
}

sub info{
    my $str=shift;
    print "[",scalar(localtime),"] $str ...\n";
}


